modules/crm_core_match/src/Form/MatcherForm.php

```diff
@@ -0,0 +1,205 @@
+        '#type' => 'item',
+        '#title' => $this->t('Matcher Plugin'),
+        '#markup' => (string) crm_core_match_matcher_manager()->getDefinition($matcher->plugin_id)['title'],
+      );
+    }
+
+    $form['plugin_container'] = array(
+      '#type' => 'container',
+      '#prefix' => '<div id="crm-core-match-match-engine-plugin">',
+      '#suffix' => '</div>',
+    );
+
+    if (isset($matcher->plugin_id) && $plugin = $matcher->getPlugin()) {
+      $form['plugin_container']['configuration'] = array(
+        '#type' => 'details',
+        '#open' => TRUE,
+        '#title' => $this->t('@plugin configuration', ['@plugin' => $matcher->getPluginTitle()]),
+        '#tree' => TRUE,
+      );
+      $form['plugin_container']['configuration'] += (array) $plugin->buildConfigurationForm($form['plugin_container']['configuration'], $form_state);
+    }
+
+    return $form;
+  }
+
+  /**
+   * Handles submit call when sensor type is selected.
+   */
+  public function submitSelectPlugin(array $form, FormStateInterface $form_state) {
+    $this->entity = $this->buildEntity($form, $form_state);
+    $form_state->setRebuild();
+  }
+
+  /**
+   * Handles switching the configuration type selector.
+   */
+  public function ajaxReplacePluginSpecificForm($form, FormStateInterface $form_state) {
+    return $form['plugin_container'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+
+    /** @var \Drupal\crm_core_match\Matcher\MatcherConfigInterface $matcher */
+    $matcher = $this->entity;
+    /** @var \Drupal\crm_core_match\Plugin\crm_core_match\engine\MatchEngineInterface $plugin */
+    if ($matcher->isNew()) {
+      $plugin_id = $form_state->getValue('plugin_id');
+      $plugin = crm_core_match_matcher_manager()->createInstance($plugin_id, array('plugin_config' => $matcher));
+    }
+    else {
+      $plugin = $matcher->getPlugin();
+    }
+
+    $plugin->validateConfigurationForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+
+    /** @var \Drupal\crm_core_match\Entity\Matcher $matcher */
+    $matcher = $this->entity;
+    $plugin = $matcher->getPlugin();
+
+    $plugin->submitConfigurationForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function save(array $form, FormStateInterface $form_state) {
+    $this->entity->save();
+    drupal_set_message($this->t('The configuration options have been saved.'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function copyFormValuesToEntity(EntityInterface $entity, array $form, FormStateInterface $form_state) {
+    parent::copyFormValuesToEntity($entity, $form, $form_state);
+
+    foreach ($form_state->getValue('configuration') as $key => $value) {
+      switch ($key) {
+        case 'rules':
```

Comment:
Moved the ```copyFormValuesToEntity()``` logic to ```submitConfigurationForm()``` method of the plugin. I guess in the followup we can implement **EntityWithPluginCollectionInterface** and simplify the things.

Btw, it would be nice to have the same form and configuration structure, so we can just save. :)